File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-service/src/test/java/io/github/crabzilla/example1/Example1EventProjectorIT.java
package io.github.crabzilla.example1;

import io.github.crabzilla.example1.customer.CustomerActivated;
import io.github.crabzilla.example1.customer.CustomerCreated;
import io.github.crabzilla.example1.customer.CustomerId;
import io.github.crabzilla.example1.customer.CustomerModule;
import io.github.crabzilla.vertx.projection.ProjectionData;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import org.jdbi.v3.core.Handle;
import org.junit.Before;
import org.junit.Test;

import java.time.Instant;
import java.util.UUID;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Assertions.assertThat;

public class Example1EventProjectorIT {

  Example1Component app ;

  @Before
  public void setUp() {

    Vertx vertx = Vertx.vertx();

    JsonObject config = new JsonObject();

    config.put("database.driver", "com.mysql.cj.jdbc.Driver");
    config.put("database.url", "jdbc:mysql://127.0.0.1:3306/example1db?serverTimezone=UTC&useSSL=false");
    config.put("database.user", "root");
    config.put("database.password", "my-secret-pwd");
    config.put("database.pool.max.size", 10);

    app = DaggerExample1Component.builder()
            .crabzillaModule(new CrabzillaModule(vertx, config))
            .customerModule(new CustomerModule(vertx, config))
            .example1Module(new Example1Module(vertx, config))
            .build();

    Handle h = app.jdbi().open();
    h.createScript("DELETE FROM units_of_work").execute();
    h.createScript("DELETE FROM customer_summary").execute();
    h.commit();

  }


  @Test
  public void canProjectTwoEvents() throws Exception {

    CustomerId id = new CustomerId("customer#1");
    CustomerCreated event1 = new CustomerCreated(id,  "customer1");
    CustomerActivated event2 = new CustomerActivated("a good reason", Instant.now());
    ProjectionData projectionData = new ProjectionData(UUID.randomUUID(), 1L, id.stringValue(), asList(event1, event2));

    app.eventProjector().handle(singletonList(projectionData));

    Handle h = app.jdbi().open();
    CustomerSummaryDao dao = h.attach(CustomerSummaryDao.class);
    CustomerSummary fromDb = dao.getAll().get(0);
    h.commit();

    assertThat(fromDb).isEqualToComparingFieldByField(new CustomerSummary(id.stringValue(), event1.getName(), true));

  }

}

File: crabzilla-kotlin-example1/crabzilla-kotlin-example1-service/src/test/java/io/github/crabzilla/example1/Example1EventProjectorIT.kt
package io.github.crabzilla.example1

import com.fasterxml.jackson.databind.ObjectMapper
import com.google.inject.Guice
import io.vertx.core.Vertx
import org.assertj.core.api.Assertions.assertThat
import io.github.crabzilla.example1.customer.CustomerActivated
import io.github.crabzilla.example1.customer.CustomerCreated
import io.github.crabzilla.example1.customer.CustomerId
import io.github.crabzilla.model.DomainEvent
import io.github.crabzilla.stack.DbConcurrencyException
import io.github.crabzilla.stack.EventProjector
import io.github.crabzilla.stack.ProjectionData
import org.jdbi.v3.core.Jdbi
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.*
import java.util.Arrays.asList
import javax.inject.Inject

@DisplayName("An Example1EventProjector")
class Example1EventProjectorIT {

  @Inject
  internal lateinit var mapper: ObjectMapper

  @Inject
  internal lateinit var jdbi: Jdbi

  @Inject
  internal lateinit var eventProjector: EventProjector<CustomerSummaryDao>

  @BeforeEach
  fun setup() {

    Guice.createInjector(Example1Module(Vertx.vertx())).injectMembers(this)

    val h = jdbi.open()
    h.createScript("DELETE FROM units_of_work").execute()
    h.createScript("DELETE FROM customer_summary").execute()
    h.commit()

  }


  @Test
  @Throws(DbConcurrencyException::class)
  fun can_project_two_events() {

    val id = CustomerId("customer#1")
    val event1 = CustomerCreated(id, "customer1")
    val event2 = CustomerActivated("a good reason", Instant.now())
    val projectionData = ProjectionData(UUID.randomUUID(), 1L, id.stringValue(), asList<DomainEvent>(event1, event2))

    eventProjector!!.handle(listOf(projectionData))

    val h = jdbi.open()
    val dao = h.attach(CustomerSummaryDao::class.java)
    val fromDb = dao.getAll()[0]
    h.commit()
    //    System.out.printf("from  db: " + fromDb);

    assertThat(fromDb).isEqualToComparingFieldByField(CustomerSummary(id.stringValue(), event1.name, true))

  }

}
